### PR TITLE
Search feature

### PR DESCRIPTION
**Pull request by nmarkovic04**

### CODE REVIEW

File: AutocompleteClient/FW/Config/AutocompleteConfig.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+//  Constants.swift
+//  Constructor.io
+//
+//  Copyright © Constructor.io. All rights reserved.
+//  http://constructor.io/
+//
+
+import Foundation
+
+public struct AutocompleteConfig {
```
Comment (rubikzube) [inline AutocompleteClient/FW/Config/AutocompleteConfig.swift]: I think we can remove this struct as it duplicates the ConstructorIO config?

File: AutocompleteClient/FW/API/Parser/Search/SearchResponseParser.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                
+                let searchFacets: [SearchResultFacet]? = (dataObj["facets"] as? [JSONObject])?.flatMap{ searchFacetObj in
+                    return SearchResultFacet(json: searchFacetObj)
+                }
+                
+                let groups: [CIOGroup]? = (dataObj["groups"] as? [JSONObject])?.flatMap({ groupObj in
+                    return CIOGroup(json: groupObj)
+                })
+                
+                return SearchResult(id: id, value: value, url: url, price: price, quantity: quantity, imageURL: imageURL, facets: searchFacets, groups: groups)
```
Comment (rubikzube) [inline AutocompleteClient/FW/API/Parser/Search/SearchResponseParser.swift]: Is the non-standard metadata available?

File: AutocompleteClient/FW/Logic/Result/Search/Facet.swift [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+//
+//  Copyright © Constructor.io. All rights reserved.
+//  http://constructor.io/
+//
+
+import UIKit
+
+struct Facet{
+    let name: String
+    let options: [FacetOption]
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/Search/Facet.swift]: Facets have four properties now, sample below

```
display_name: "Age Range"
name: "Age Range"
options: []
type: "multiple"
```

File: AutocompleteClient/FW/Logic/Result/Search/FacetOption.swift [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+//  AutocompleteClient
+//
+//  Copyright © Constructor.io. All rights reserved.
+//  http://constructor.io/
+//
+
+import UIKit
+
+struct FacetOption{
+    let count: Int
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/Search/FacetOption.swift]: Facets have five properties now, sample below
```
count: 74
data: {}
display_name: "20s"
status: "selected"
value: "20s"
```

File: AutocompleteClient/FW/Logic/Result/Search/SearchResult.swift [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+struct SearchResult {
+    let id: String
+    let value: String
+    let url: String?
+    let price: String?
+    let quantity: String?
+    let imageURL: String?
+    let facets: [SearchResultFacet]?
+    let groups: [CIOGroup]?
+
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/Search/SearchResult.swift]: This data structure isn't quite right.  A search result will have three properties:  
```
value: string
data: object
matched_terms: string array
```

and the data property will have:
```
description: string
id: string
url: string
image_url: string
groups: object array
facets: object array
metadata: object
```

File: AutocompleteClient/FW/Logic/Result/Search/SearchResultFacet.swift [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+//  SearchResultFacet.swift
+//  AutocompleteClient
+//
+//  Copyright © Constructor.io. All rights reserved.
+//  http://constructor.io/
+//
+
+import Foundation
+
+struct SearchResultFacet{
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/Search/SearchResultFacet.swift]: Let's remove this for the moment.

File: AutocompleteClient/FW/Logic/Result/Search/CIOSearchResponse.swift [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+import Foundation
+
+/**
+ Struct representing the search data response from the server.
+ */
+public struct CIOSearchResponse {
+    let facets: [Facet]
+    let results: [SearchResult]
+    let redirectInfo: SearchRedirectInfo?
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/Search/CIOSearchResponse.swift]: Need the following extra properties:
```
groups,
sort options,
total number results
result id
```

File: AutocompleteClient/FW/Config/AutocompleteConfig.swift (hunk 1 above)
Comment (nmarkovic04) [reply AutocompleteClient/FW/Config/AutocompleteConfig.swift]: Sure. This file isn't even in the project, but was picked up by git. Something probably happened during rebase.

File: AutocompleteClient/Constants/Constants.swift [hunk 8]
```diff
@@ -77,6 +77,15 @@ public struct Constants {
         public static let defaultItemCountPerSection = 10
     }
 
+    public struct SearchQuery {
+        public static let pathString = "search"
+        public static let page = "page"
+        public static let groupFilter = "filters[group_id]"
+        public static let facetFilterKey = { (key: String) -> String in "filters[\(key)]" }
+        public static let numResultsPerPage = "num_results_per_page"
+        public static let numResultsPerPageForSectionNamePrefix = "num_results_per_page_"
```
Comment (rubikzube) [inline AutocompleteClient/Constants/Constants.swift]: This parameter doesn't exist for search results... we can only have one section which defaults to Products, `section=Products`

File: AutocompleteClient/FW/Logic/Request/Builder/RequestBuilder+QueryItems.swift [hunk 9]
```diff
@@ -69,6 +69,37 @@ extension RequestBuilder {
...
+        queryItems.add(URLQueryItem(name: Constants.SearchQuery.groupFilter, value: filter))
+    }
+    
+    public func set(facetFilters: [Filter]?){
+        guard let filters = facetFilters else { return }
+        for filter in filters{
+            self.set(facetFilter: filter)
+        }
+    }
+
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Request/Builder/RequestBuilder+QueryItems.swift]: Prolly need `setSection` added here

File: AutocompleteClient/FW/Logic/Request/CIOAutocompleteQuery.swift [hunk 10]
```diff
@@ -32,3 +32,34 @@ public struct CIOAutocompleteQuery: CIORequestData {
...
     }
 }
+
+public struct CIOSearchQuery: CIORequestData{
+
+    let query: String
+    let filters: SearchFilters?
+    let page: Int
+    let numResultsPerPage: Int
+    let numResultsPerPageForSection: [String: Int]?
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Request/CIOAutocompleteQuery.swift]: Remove `numResultsPerPageForSection`, add `section`

File: AutocompleteClient/FW/Logic/Result/Search/CIOSearchResponse.swift (hunk 7 above)
Comment (rubikzube) [reply AutocompleteClient/FW/Logic/Result/Search/CIOSearchResponse.swift]: Still need these fields

**Pull request merged**